File: src/main/java/com/github/jactor/web/dto/UserModel.java
package com.github.jactor.web.dto;

import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class UserModel {

  private final AddressDto address;
  private final PersonDto person;
  private final UserDto user;

  public UserModel(UserDto user) {
    this.user = user;
    person = Optional.ofNullable(user).map(UserDto::getPerson).orElse(null);
    address = Optional.ofNullable(person).map(PersonDto::getAddress).orElse(null);
  }

  public List<String> fetchAddress() {
    return Stream.of(fetchAddressLine1(), fetchAddressLine2(), fetchAddressLine3(), fetchZipCode(), fetchCity())
        .filter(Objects::nonNull)
        .collect(Collectors.toList());
  }

  public String fetchFullName() {
    String surname = fetchSurname();

    return fetchFirstname()
        .map(s -> s + " " + surname)
        .orElse(surname);
  }

  private String fetchAddressLine1() {
    return Optional.ofNullable(address)
        .map(AddressDto::getAddressLine1).orElse(null);
  }

  private String fetchAddressLine2() {
    return Optional.ofNullable(address)
        .map(AddressDto::getAddressLine2).orElse(null);
  }

  private String fetchAddressLine3() {
    return Optional.ofNullable(address)
        .map(AddressDto::getAddressLine3).orElse(null);
  }

  private String fetchCity() {
    return Optional.ofNullable(address)
        .map(AddressDto::getCity).orElse(null);
  }

  private String fetchZipCode() {
    return String.valueOf(
        Optional.ofNullable(address)
            .map(AddressDto::getZipCode)
            .orElse(null)
    );
  }

  private Optional<String> fetchFirstname() {
    return Optional.ofNullable(person)
        .map(PersonDto::getFirstName);
  }

  private String fetchSurname() {
    return Optional.ofNullable(person)
        .map(PersonDto::getSurname).orElse("");
  }

  public String fetchUsername() {
    return user.getUsername();
  }

  Optional<String> fetchDescriptionCode() {
    return Optional.ofNullable(person)
        .map(PersonDto::getDescription);
  }
}

File: web/src/main/kotlin/com/github/jactor/web/dto/UserModel.kt
package com.github.jactor.web.dto

import com.github.jactor.shared.dto.AddressDto
import com.github.jactor.shared.dto.PersonDto
import com.github.jactor.shared.dto.UserDto
import java.util.Optional

class UserModel(private val address: AddressDto?, private val person: PersonDto?, private val user: UserDto) {
    constructor(user: UserDto) : this(address = user.person?.address, person = user.person, user = user)

    fun fetchAddress(): List<String> {
        return listOf(
            address?.addressLine1 ?: "",
            address?.addressLine2 ?: "",
            address?.addressLine3 ?: "",
            address?.zipCode ?: "",
            address?.city ?: ""
        )
    }

    fun fetchFullName(): String {
        val firstName = person?.firstName

        return if (firstName != null) "$firstName ${person?.surname}" else person?.surname ?: ""
    }

    fun fetchUsername() = user.username ?: ""
    fun fetchDescriptionCode() = Optional.ofNullable(person?.description)
}
